### PR TITLE
Setting up infrastructure for UTF8StringExperiment in order to have full ci and official build support

### PR DESCRIPTION
**Pull request by joperezr**
This isn't ready for merge, I will be adding several commits in order to identify all infra that will be required in order to get both CI and official build support for this first experiment in runtimelab.

cc: @safern to take a look on yml changes. Worth noting that runtimelab.yml is a new yml which the main pipeline of this repo will be based on. runtimelab.yml will be checked into dotnet/runtime when we feel it is at a ready enough state for experiments on this repo.

FYI: @jkotas

### CODE REVIEW

File: eng/pipelines/runtimelab.yml [hunk 1]
```diff
@@ -51,3 +51,33 @@ jobs:
...
+  parameters:
+    jobTemplate: /eng/pipelines/common/global-build-job.yml
+    buildConfig: release
+    platforms:
+    - Linux_x64
+    - Windows_NT_x64
+    jobParameters:
+      testGroup: innerloop
+      nameSuffix: Runtime_Debug
+      buildArgs: -s clr+libs+installer -c debug -runtimeConfiguration release
```
Comment (jkotas) [inline eng/pipelines/runtimelab.yml:68]:
```suggestion
      buildArgs: -s clr+libs+installer -c debug -runtimeConfiguration checked
```
to get better coverage.

Comment (joperezr) [reply eng/pipelines/runtimelab.yml:68]: I can add a checked leg but I assume libraries tests should always use release coreclr right?

Comment (jkotas) [reply eng/pipelines/runtimelab.yml:68]: The Utf8String implementation is in CoreLib and all tests for Utf8String are under libraries.

If we were to always use release coreclr with libraries tests, we would get no coverage for debug version of Utf8String, e.g. the asserts in Utf8String would be never exercised.

I think we should have two legs: release everything and debug/checked everything.

Comment (joperezr) [reply eng/pipelines/runtimelab.yml]: Done.

File: eng/pipelines/runtimelab/runtimelab-post-build-steps.yml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+      tarCompression: $(tarCompression)
+      includeRootFolder: false
+      archiveExtension: $(archiveExtension)
+      artifactName: CoreCLRProduct__$(osGroup)$(osSubgroup)_$(archType)_$(buildConfig)
+      displayName: 'CoreCLR product build'
+
+  # Zip Libraries Build Output
+  - template: /eng/pipelines/common/upload-artifact-step.yml
+    parameters:
+      rootFolder: $(Build.ArtifactStagingDirectory)/artifacts
```
Comment (safern) [inline eng/pipelines/runtimelab/runtimelab-post-build-steps.yml:16]: This will include all artifacts, coreclr, and installer as well, which is pretty big. I suggest you copying what you need for running tests:
For example we do this in runtime: https://github.com/dotnet/runtime/blob/master/eng/pipelines/libraries/build-job.yml#L101-L123

That will save build time and avoid being throttled by azdo, etc.

File: eng/pipelines/runtimelab.yml [hunk 3]
```diff
@@ -53,31 +53,33 @@ jobs:
...
     jobTemplate: /eng/pipelines/common/global-build-job.yml
-    buildConfig: debug
+    buildConfig: release
     platforms:
     - Linux_x64
     - Windows_NT_x64
     jobParameters:
       testGroup: innerloop
       nameSuffix: Runtime_Release
       buildArgs: -s clr+libs+installer -c release -runtimeConfiguration release
```
Comment (safern) [inline eng/pipelines/runtimelab.yml:84]: You don't need to pass down `-runtimeConfiguration` here, right?

File: eng/pipelines/common/global-build-job.yml [hunk 4]
```diff
@@ -22,7 +24,25 @@ jobs:
     workspace:
       clean: all
 
-    variables: ${{ parameters.variables }}
+    variables:
+    - name: osGroup
```
Comment (safern) [inline eng/pipelines/common/global-build-job.yml:28]: Do these variables have to be named like that? I've seen problems where we somehow matched the MSBuild property name and since AzDo sets these variables as env vars, then you hit weird errors. Maybe append them with `_` or something to avoid clashes in between MSBuild and yml variables.

Comment (safern) [reply eng/pipelines/common/global-build-job.yml:28]: Also it is kind of weird to define these variables here just to make extraStepTemplates of runtime lab work. 

I believe osGroup, osSubgroup, archType and buildConfig should be taken as parameters by extra step template. 

buildConfigUpper is already used in a couple of places and defined in `mono/xplat-job` and `coreclr/xplat-job`. 
https://github.com/dotnet/runtime/search?l=YAML&q=buildConfigUpper

I think we should instead remove it from there and move the variable declaration to `xplat-setup.yml` which is common to all jobs, that way we ensure it is always defined, no way to duplicate logic across multiple ymls.

File: eng/pipelines/runtimelab/runtimelab-post-build-steps.yml (hunk 2 above)
Comment (joperezr) [reply eng/pipelines/runtimelab/runtimelab-post-build-steps.yml:16]: good point, actually I copied this part from runtime so I'm actually using ArtifactStagingDirectory as well, the problem is that I didn't prepare that directory with the things I'm interested in which is why the build is failing, I'll fix this.

File: eng/pipelines/common/templates/runtimes/run-test-job.yml [hunk 5]
```diff
@@ -20,6 +20,8 @@ parameters:
   pool: ''
   runtimeFlavor: 'coreclr'
   runtimeFlavorDisplayName: 'CoreCLR'
+  overrideDependsOn: false
```
Comment (safern) [inline eng/pipelines/common/templates/runtimes/run-test-job.yml:23]: We don't need this extra parameter. We could just test if `dependsOn` has any value, i.e:
```yml
- ${{ if ne(parameters.dependsOn[0], '') }}:
```

File: eng/pipelines/common/templates/runtimes/build-test-job.yml [hunk 6]
```diff
@@ -23,6 +23,9 @@ parameters:
   runtimeFlavorDisplayName: 'CoreCLR'
   runtimeVariant: ''
 
+  overrideDependsOn: false
```
Comment (safern) [inline eng/pipelines/common/templates/runtimes/build-test-job.yml:26]: We can remove this parameter as well.

File: eng/pipelines/libraries/build-test-job.yml [hunk 7]
```diff
@@ -16,6 +16,8 @@ parameters:
   variables: {}
   condition: true
   runTests: false
+  overrideDependsOn: false
```
Comment (safern) [inline eng/pipelines/libraries/build-test-job.yml:19]: We don't need this parameter either.

File: eng/pipelines/libraries/prepare-publish-for-tests.yml [hunk 8]
```diff
@@ -0,0 +1,26 @@
+# Steps used to prepare the Artifacts Staging Directory with required files for
```
Comment (safern) [inline eng/pipelines/libraries/prepare-publish-for-tests.yml:1]: I think it would be better to have these 2 templates as 1 template, that will make it easier to read what we're already copying and so that we don't duplicate any of these steps.

File: eng/pipelines/libraries/run-test-job.yml [hunk 9]
```diff
@@ -22,6 +22,8 @@ parameters:
   # stress modes that each test will be run with. This is the same usage as 'testGroup' in
   # eng/pipelines/common/templates/runtimes/run-test-job.yml.
   coreclrTestGroup: ''
+  overrideDependsOn: false
```
Comment (safern) [inline eng/pipelines/libraries/run-test-job.yml:25]: We don't need this either.

File: eng/pipelines/runtimelab.yml [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+    jobTemplate: /eng/pipelines/common/templates/runtimes/build-test-job.yml
+    buildConfig: checked
+    platforms:
+    - Linux_x64
+    jobParameters:
+      testGroup: innerloop
+      liveLibrariesBuildConfig: Release
+      overrideDependsOn: true
+      dependsOn:
+      - build_Linux_x64_checked_
```
Comment (safern) [inline eng/pipelines/runtimelab.yml:113]: Why is one upper-case and one lower-case? And why is there a trailing `_` there?

File: eng/pipelines/runtimelab.yml [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+      isOfficialBuild: false
+      liveRuntimeBuildConfig: Release
+      testScope: innerloop
+      overrideDependsOn: true
+      dependsOn:
+      - build_Windows_NT_x64_Release_
+
+#
+# Libraries Release Test Execution against a release coreclr runtime
+# Only when the PR contains a libraries change
```
Comment (safern) [inline eng/pipelines/runtimelab.yml:184]: We can remove the comments that mention, `Only when the PR contains a ....`

File: eng/pipelines/runtimelab.yml [hunk 12]
```diff
@@ -0,0 +1,220 @@
+# Setting batch to true, triggers one build at a time.
```
Comment (safern) [inline eng/pipelines/runtimelab.yml:1]: should we move this template to `eng/pipelines/runtimelab/runtimelab.yml`? so that when we grep for parameters that are only used here (like `dependsOn`), we find them in `runtime` and no-one breaks the `runtimelab` pipeline?

File: eng/pipelines/runtimelab.yml (hunk 10 above)
Comment (joperezr) [reply eng/pipelines/runtimelab.yml:113]: lowercase was because I was using lowercase for checked build config above which I'll fix. The trailing underscore is the default name because we have a support for a build name suffix which I'm not using.

File: eng/pipelines/runtimelab.yml (hunk 12 above)
Comment (joperezr) [reply eng/pipelines/runtimelab.yml:1]: not sure I follow the protecting against the break comment but ideally I would like to keep it in the root of pipelines if possible since at least in my mind all the ymls in the root of pipelines represent actual pipelines we have in AzDO. I don't have a strong opinion on this though, so if you still want me to move it I'm fine with that.

Comment (safern) [reply eng/pipelines/runtimelab.yml:1]: The thing in question here, is if this yml will be ported to `dotnet/runtime`? I just want it to be there, so that when we grep for yml parameters, or templates usage, we get usages on this template.

File: eng/pipelines/libraries/prepare-publish-for-tests.yml [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
```
Comment (safern) [inline eng/pipelines/libraries/prepare-publish-for-tests.yml:1]: Nit: maybe name this template, `prepare-publish-for-framework` or something like that? Cause this is not just for tests, this is also for the official build.

File: eng/pipelines/runtimelab.yml [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+    - Linux_x64
+    helixQueuesTemplate: /eng/pipelines/libraries/helix-queues-setup.yml
+    jobParameters:
+      isFullMatrix: false
+      isOfficialBuild: false
+      testScope: innerloop
+      liveRuntimeBuildConfig: Release
+      dependsOnTestBuildConfiguration: Release
+      dependsOnTestArchitecture: x64
+      overrideDependsOn: true
```
Comment (safern) [inline eng/pipelines/runtimelab.yml:198]: Remove all the not needed `overrideDependsOn` from this file.

File: eng/pipelines/runtimelab.yml (hunk 10 above)
Comment (safern) [reply eng/pipelines/runtimelab.yml]: I see makes sense.

File: eng/pipelines/runtimelab.yml (hunk 12 above)
Comment (joperezr) [reply eng/pipelines/runtimelab.yml:1]: Yes the idea is to port this guy to dotnet/runtime as well. The reason for that is that we want experiments in runtimelab to only contain the diff of the actual feature being added, as opposed to change in ymls and such other files that would make it harder to maintain or to port back to runtime once the experiment is ready for that.

Comment (safern) [reply eng/pipelines/runtimelab.yml:1]: Ok, cool then I'm fine with where it is.

**Comment by joperezr** [comment]
@jkotas @safern Do you guys have anything to add to CI here or is this ready to go? we are currently running libraries and coreclr tests, you can probably check the checks to see what's included for now.

**Comment by joperezr** [comment]
Actually this will be broken by latest changes @safern made on dotnet/runtime so let me do one more pull of dotnet/runtime here and adjust that.

**Comment by joperezr** [comment]
Ok, now I have addressed that conflict and CI is running now. Let me know if you guys want me to add anything else but if not I'll go ahead and merge once this is green in order to start working on porting this back to dotnet/runtime and also to enable official build support for runtimelab.

File: eng/pipelines/libraries/build-job.yml [hunk 15]
```diff
@@ -93,56 +93,16 @@ jobs:
...
+              isOfficialBuild: ${{ parameters.isOfficialBuild }}
+          
           - task: CopyFiles@2
             displayName: Prepare docs folder to publish
             inputs:
               sourceFolder: $(Build.SourcesDirectory)/artifacts/bin/docs
               targetFolder: $(Build.ArtifactStagingDirectory)/artifacts/bin/docs
 
-          - task: CopyFiles@2
-            displayName: Prepare native folder to publish
```
Comment (safern) [inline eng/pipelines/libraries/build-job.yml:134]: Just curious, why is this needed in the runtimelab setup?

Comment (safern) [reply eng/pipelines/libraries/build-job.yml:134]: I guess it will be needed for the official build, nevermind.

File: eng/pipelines/libraries/build-job.yml [hunk 16]
```diff
@@ -93,56 +93,16 @@ jobs:
...
-            displayName: Prepare shared framework runtime folder to publish
-            inputs:
-              sourceFolder: $(Build.SourcesDirectory)/artifacts/bin/pkg
-              targetFolder: $(Build.ArtifactStagingDirectory)/artifacts/bin/pkg
-
+          - template: /eng/pipelines/libraries/prepare-publish-for-tests.yml
+            parameters:
+              isOfficialBuild: ${{ parameters.isOfficialBuild }}
+          
           - task: CopyFiles@2
```
Comment (safern) [inline eng/pipelines/libraries/build-job.yml:100]: The docs should also be needed for the official build to produce packs with docs?

File: eng/pipelines/runtimelab.yml [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+- template: /eng/pipelines/common/platform-matrix.yml
+  parameters:
+    jobTemplate: /eng/pipelines/common/global-build-job.yml
+    buildConfig: release
+    platforms:
+    - Windows_NT_x64
+    jobParameters:
+      testGroup: innerloop
+      nameSuffix: All_Configurations
+      buildArgs: -s clr.corelib+libs -c $(_BuildConfig) -allConfigurations
```
Comment (safern) [inline eng/pipelines/runtimelab.yml:97]: Nit: you no longer need `clr.corelib` as it is now a `P2P` from libraries projects instead of `ReferenceFromRuntime`

File: eng/pipelines/runtimelab.yml [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+    helixQueuesTemplate: /eng/pipelines/libraries/helix-queues-setup.yml
+    jobParameters:
+      isFullMatrix: false
+      isOfficialBuild: false
+      testScope: innerloop
+      liveRuntimeBuildConfig: Release
+      dependsOnTestBuildConfiguration: Release
+      dependsOnTestArchitecture: x64
+      dependsOn:
+      - build_Windows_NT_x64_Release_
```
Comment (safern) [inline eng/pipelines/runtimelab.yml:183]: This will add a lot of duplication if we want to build for more platforms where the only thing that differs is the `dependsOn`. I guess we could be smarter later on and have a prefix and then use `osGroup`, `archType`, `buildConfig` and a `suffix` and then construct the dependsOn on the job template directly.

Don't have to address this on the PR but just a note that it might make the entry point yml "cleaner"

File: src/libraries/libraries-packages.proj [hunk 19]
```diff
@@ -18,6 +18,7 @@
     <ProjectReference Include="$(PkgDir)*\*.proj" Exclude="$(PkgDir)test\*" />
     <ProjectReference Include="$(MSBuildThisFileDirectory)*\pkg\**\*.pkgproj" Condition="('$(BuildAllConfigurations)' == 'true' or '$(DotNetBuildFromSource)' == 'true') And '$(BuildAllOOBPackages)' == 'true'" />
     <!-- If setting BuildAllOOBPackages to false, add bellow the individual OOB packages you want to continue to build -->
```
Comment (safern) [inline src/libraries/libraries-packages.proj:20]: Where is `BuildAllOOBPackages` set?

File: eng/pipelines/libraries/build-job.yml (hunk 16 above)
Comment (joperezr) [reply eng/pipelines/libraries/build-job.yml:100]: We don't yet build the product build and the reason why I removed this is because I don't currently produce the docs folder with the build command so this fails unless I start producing them.

File: eng/pipelines/runtimelab.yml (hunk 17 above)
Comment (joperezr) [reply eng/pipelines/runtimelab.yml:97]: Noted, will remove

File: eng/pipelines/runtimelab.yml (hunk 18 above)
Comment (joperezr) [reply eng/pipelines/runtimelab.yml:183]: I agree, I could also just keep a single one but then would have to wait for build that aren't really required in order to start running other jobs.

File: src/libraries/libraries-packages.proj (hunk 19 above)
Comment (joperezr) [reply src/libraries/libraries-packages.proj:20]: A few lines above this one on this same file:

https://github.com/dotnet/runtimelab/blob/82c3b4c7fa34091524d21fe4ed8977c2da893570/src/libraries/libraries-packages.proj#L7

File: eng/pipelines/libraries/build-job.yml (hunk 16 above)
Comment (safern) [reply eng/pipelines/libraries/build-job.yml:100]: The reason why I think we should keep `prepare-publish-for-tests` on pair with what we need on runtime and runtimelab is to avoid missing something or adding something to `prepare-publish-for-tests` that was in `build-job` because we didn't see it there. It is just weird to be in a `hybrid` mode of some things are in the shared template and some are not. You will eventually need to publish these, so why not put them there?

You could also put it there and disable via a parameter that the template takes.

File: eng/pipelines/runtimelab/runtimelab-post-build-steps.yml [hunk 20]
```diff
@@ -23,15 +24,17 @@ steps:
...
-    parameters:
-      rootFolder: $(Build.SourcesDirectory)/artifacts/helix
-      includeRootFolder: true
-      archiveType: $(archiveType)
-      archiveExtension: $(archiveExtension)
-      tarCompression: $(tarCompression)
-      artifactName: libraries_test_assets_${{ parameters.osGroup }}_${{ parameters.archType }}_${{ parameters.buildConfig }}
-      displayName: Test Assets
+  # Zip Test Build
+  - ${{ if eq(parameters.uploadTests, 'true') }}:
```
Comment (safern) [inline eng/pipelines/runtimelab/runtimelab-post-build-steps.yml:28]: Default the parameter to false, and do a boolean comparison, it is usually better to  get precise results, comparing against a string could mess up the build based on casing or other evaluations that azdo does.

**Comment by joperezr** [comment]
For some reason, the badge hasn't updated here but PR is all green now. Thanks all for the review; I will now follow up and make a PR with these changes back to dotnet/runtime.
